### PR TITLE
[Core][Modes] New buffer object to track underlying mode allocations …

### PR DESCRIPTION
**Pull request by noelchalmers**
## Description

This PR slightly changes how reference to backend memory allocations are counted so they can be more reliably destroyed, and prevent memory access violations via invalidated pointers. 

The changes are entirely internal to OCCA and do not change the public API, or alter current behavior of OCCA applications.

#### Summary of issue: 
Currently `modeMemory_t` objects do not track whether they were created by slicing another `modeMemory_t` object. Consequently, only the original `modeMemory_t` object can free the underlying device memory allocation, and will do so even if there still exist `modeMemory_t` objects referencing the device memory. This results in dangling references.

E.g. 

```c++
{
  occa::memory o_a;
  {
    occa::memory o_A = device.malloc<float>(2*entries); //Create a new buffer
    o_a = o_A + entries; //  o_a is a slice into o_A
  } //o_A is destroyed here because it's reference counting is separate from o_a.
    // The underlying device memory is also destroyed

  // o_a is now invalid, but doesn't know the original device memory is free'd

  // Using o_a will result in an access violation
  myKernel(entries, o_a); //Fault

} //o_a destroyed here, but since it's not the origin, there's no double free
``` 

This PR addresses this behavior. The new `buffer` objects wrap device memory allocations and track their `modeMemory_t` references, created via slicing. The `buffer` and wrapped device allocation are destroyed once their reference count reaches zero.

### CODE REVIEW

File: src/occa/internal/modes/cuda/memory.cpp [hunk 1]
```diff
@@ -1,56 +1,45 @@
...
+      isUnified = b->isUnified;
+      useHostPtr = b->useHostPtr;
+      if (useHostPtr) {
+        ptr = b->ptr + offset;
+      } else if (isUnified) {
+        ptr = b->ptr + offset;
+        cuPtr = b->cuPtr + offset;
+      } else {
+        cuPtr = b->cuPtr + offset;
+      }
```
Comment (dmed256) [inline src/occa/internal/modes/cuda/memory.cpp:21]: Nit:
```suggestion
      if (isUnified || useHostPtr) {
        ptr = b->ptr + offset;
      }
      if (isUnified || !useHostPtr) {
        cuPtr = b->cuPtr + offset;
      }
```

File: src/occa/internal/modes/dpcpp/buffer.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      occa::modeBuffer_t(modeDevice_, size_, properties_) {}
+
+    buffer::~buffer() {
+      if (!isWrapped) {
+        if (ptr) {
+          auto& dpcpp_device = getDpcppDevice(modeDevice);
+          OCCA_DPCPP_ERROR("Memory: Freeing SYCL alloc'd memory",
+                           ::sycl::free(ptr,dpcpp_device.dpcppContext));
+        }
+      }
```
Comment (dmed256) [inline src/occa/internal/modes/dpcpp/buffer.cpp:22]: Nit:
```suggestion
      if (!isWrapped && ptr) {
        auto& dpcpp_device = getDpcppDevice(modeDevice);
        OCCA_DPCPP_ERROR("Memory: Freeing SYCL alloc'd memory",
                         ::sycl::free(ptr,dpcpp_device.dpcppContext));
      }
```

File: src/occa/internal/modes/dpcpp/device.cpp [hunk 3]
```diff
@@ -322,58 +323,16 @@ namespace occa
...
-      OCCA_ERROR("DPCPP: malloc_shared failed!", nullptr != mem->ptr);
-
-      if (nullptr != src)
-      {
-        // Since memory is automatically migrated, we only need to copy
-        // between host pointers.
-        ::memcpy(mem->ptr, src, bytes);
-      }
+      if (src != NULL)
+        mem->copyFrom(src, bytes, 0, props);
```
Comment (dmed256) [inline src/occa/internal/modes/dpcpp/device.cpp:335]:
```suggestion
      if (src != NULL) {
        mem->copyFrom(src, bytes, 0, props);
      }
```

File: src/occa/internal/modes/metal/device.cpp [hunk 4]
```diff
@@ -275,23 +276,27 @@ namespace occa {
...
+      //create allocation
+      buffer *buf = new metal::buffer(this, bytes, props);
+      buf->malloc(bytes);
 
-      mem->metalBuffer = metalDevice.malloc(bytes, src);
+      //create slice
+      memory *mem = new metal::memory(buf, bytes, 0);
+
+      if (src)
+        mem->copyFrom(src, bytes, 0, props);
```
Comment (dmed256) [inline src/occa/internal/modes/metal/device.cpp:287]:
```suggestion
      if (src) {
        mem->copyFrom(src, bytes, 0, props);
      }
```

File: src/occa/internal/modes/opencl/device.cpp [hunk 5]
```diff
@@ -315,83 +316,28 @@ namespace occa {
...
-                                            0, bytes,
-                                            0, NULL, NULL,
-                                            &error);
-
-      OCCA_OPENCL_ERROR("Device: clEnqueueMapBuffer", error);
-
-      // Sync memory mapping
-      finish();
+      if (src)
+        mem->copyFrom(src, bytes, 0, props);
```
Comment (dmed256) [inline src/occa/internal/modes/opencl/device.cpp:328]:
```suggestion
      if (src) {
        mem->copyFrom(src, bytes, 0, props);
      }
```

File: src/occa/internal/modes/serial/device.cpp [hunk 6]
```diff
@@ -426,32 +427,33 @@ namespace occa {
...
-          ::memcpy(mem->ptr, src, bytes);
-        }
+        buf->malloc(bytes);
       }
 
+      //create slice
+      memory *mem = new serial::memory(buf, bytes, 0);
+
+      if (src && !props.get("use_host_pointer", false))
+        mem->copyFrom(src, bytes, 0, props);
```
Comment (dmed256) [inline src/occa/internal/modes/serial/device.cpp:443]:
```suggestion
      if (src && !props.get("use_host_pointer", false)) {
        mem->copyFrom(src, bytes, 0, props);
      }
```

File: src/occa/internal/core/buffer.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      removeModeMemoryRef(mem);
+      mem->modeBuffer = NULL;
+
+      delete mem;
+    }
+
+    // Remove ref from device
+    if (modeDevice) {
+      if (!isWrapped)
+        modeDevice->bytesAllocated -= size;
```
Comment (dmed256) [inline src/occa/internal/core/buffer.cpp:31]:
```suggestion
      if (!isWrapped) {
        modeDevice->bytesAllocated -= size;
      }
```

**Review by dmed256 — COMMENTED** [review]
This PR is freaking great, glad we're making memory slices safer ❤️ 
I have a few comments but overall it looks great!

**Review by dmed256 — APPROVED** [review]

File: src/core/device.cpp [hunk 8]
```diff
@@ -495,7 +495,7 @@ namespace occa {
     memory mem(modeDevice->wrapMemory(ptr, bytes, memProps));
 
     mem.setDtype(dtype);
-    mem.dontUseRefs();
+    // mem.dontUseRefs();
```
Comment (noelchalmers) [inline src/core/device.cpp:498]:
```suggestion
```

**Pull request merged**